### PR TITLE
WIP: compiler modules

### PR DESCRIPTION
**Pull request by adriaanm**
Enforce consistency of pack, docs, osgi, maven tasks.

Instead of scattering properties all over invocations of
staged-docs, staged-pack, make-bundle, copy-bundle and mvn-package,
these are now all fully determined by the project they act on.

The varying arguments to these macros are all centralized.
They are named like `project.prop`, where `prop` is one of:
description, package, dir, name, namesuffix, version,
targetdir, targetjar, jar, docroot, skipPackage, srcdir, src, srcjar.

The defaults for these properties are computed using `init-project-prop`.

### CODE REVIEW

**Comment by adriaanm** [comment]
the last commit is work in progress -- the first ones simplify the build to make the last one as small as possible

todo:
- somehow jline breaks in sbt console
- publish virtual (and how to convince the maven task to publish without a jar?) maven artifact scala-compiler that depends on -core, -doc, and -interactive? no more scala-compiler.jar? scala-compiler-*.jar instead?
- similarly for osgi

**Comment by retronym** [comment]
> publish virtual (and how to convince the maven task to publish without a jar?) maven artifact scala-compiler that depends on -core, -doc, and -interactive? no more scala-compiler.jar? scala-compiler-*.jar instead?

From command line maven, I think it is enough to have a pom with `<packaging>pom</packaging>`. The `<maven>` ant task should have the same semantics, if you can't convince the more fine-grained ant tasks to do this.

**Comment by retronym** [comment]
I guess you already know about this one, but here is an example: https://github.com/scala/scala-library-all

**Comment by skyluc** [comment]
The last commit breaks the sbinary build https://scala-webapps.epfl.ch/jenkins/job/pr-scala-integrate-ide/3608/console

**Comment by adriaanm** [comment]
Yep, I was trying to mimic scala-library-all, but the maven ant task does not let me publish without a jar.

**Comment by adriaanm** [comment]
Thanks, @skyluc! I'm abusing jenkins to speed up my integration testing as I experiment with different poms/bnds for the scala-compiler split into scala-compiler-core, scala-compiler-doc, scala-compiler-interactive (and next up: scala-compiler-repl)

**Comment by adriaanm** [comment]
Back to the original approach: slim down scala-compiler jar, moving those classes to newly introduced jars for scaladoc, presentation compiler.

**Comment by adriaanm** [comment]
@skyluc, @dragos: could you have a look at the changes to the *.bnd files? thanks!

**Comment by adriaanm** [comment]
btw, you can compile and publish this locally pretty quickly with:

```
ant -Dmaven.version.number=2.11.0-local -Dstarr.version=2.11.0-M6 -Dstarr.use.released=1 -Dlocker.skip=1 -Ddocs.skip=1 publish.local
```

File: src/build/bnd/scala-compiler.bnd [hunk 1]
```diff
@@ -6,3 +6,5 @@ Export-Package: *;version=${ver}
 Import-Package: jline.*;resolution:=optional, \
                 org.apache.tools.ant.*;resolution:=optional, \
                 *
+Require-Bundle: org.scala-lang.modules.scala-compiler-interactive_@SCALA_BINARY_VERSION@;bundle-version=@SCALA_COMPILER_INTERACTIVE_VERSION@, \
```
Comment (dragos) [inline src/build/bnd/scala-compiler.bnd:9]: Why do you need these dependencies? I thought they go in the opposite direction.

Comment (adriaanm) [inline src/build/bnd/scala-compiler.bnd:9]: We don't need them -- I was thinking this way you wouldn't have to change the IDE build.

**Comment by adriaanm** [comment]
Closing for rebasing and dbuild upgrade. Please pardon our dust.

**Pull request closed without merging**